### PR TITLE
[HLSL] Treat `main` as any other function

### LINKED ISSUE #110517: [HLSL] `main` is just a function
HLSL gives no special meaning to the function named `main`. It does give special handling to _entry points_, and the _active entry point_ when building a freestanding shader. Those cases are handled by the HLSLShader attribute, which can be implicitly applied to `main` if no other entry is specified, rather than the function name.

### PR DESCRIPTION
**Pull request by llvm-beanz**
HLSL doesn't distinguish `main` from any other function. It does treat entry points special, but they're not required to be called `main` so we have a different attribute annotation to mark them.

At the moment this change really just changes the mangling of functions named `main` in the Itanium mangling.

Fixes #110517

### CODE REVIEW

**Review by bogner — APPROVED** [review]
There is still logic elsewhere that sets the default entry point to "main", right? This looks right to me.

**Review by s-perron — APPROVED** [review]
This fixes the problem in #108567

File: clang/test/Options/enable_16bit_types_validation_spirv.hlsl [hunk 1]
```diff
@@ -4,7 +4,7 @@
 // SPIRV: error: '-fnative-half-type' option requires target HLSL Version >= 2018, but HLSL Version is 'hlsl2016'
 
 // valid: "spirv-unknown-vulkan-library"
-// valid: define spir_func void @main() #0 {
+// valid: define spir_func void @{{.*}}main{{.*}}() #0 {
```
Comment (farzonl) [inline clang/test/Options/enable_16bit_types_validation_spirv.hlsl:7]:
```suggestion
valid: define spir_func void @{{.*main.*}}() #0 {
```

File: clang/test/CodeGenHLSL/builtins/wave_get_lane_index_do_while.hlsl [hunk 2]
```diff
@@ -1,7 +1,7 @@
 // RUN: %clang_cc1 -std=hlsl2021 -finclude-default-header -x hlsl -triple \
 // RUN:   spirv-pc-vulkan-library %s -emit-llvm -disable-llvm-passes -o - | FileCheck %s
 
-// CHECK: define spir_func void @main() [[A0:#[0-9]+]] {
+// CHECK: define spir_func void @{{.*}}main{{.*}}() [[A0:#[0-9]+]] {
```
Comment (farzonl) [inline clang/test/CodeGenHLSL/builtins/wave_get_lane_index_do_while.hlsl:4]:
```suggestion
// CHECK: define spir_func void @{{.*main.*}}() [[A0:#[0-9]+]] {
```

**Review by farzonl — APPROVED** [review]

**Comment by llvm-beanz** [comment]
> There is still logic elsewhere that sets the default entry point to "main", right? This looks right to me.

Yes, and we apply the HLSLShader attribute to main and key generating the entry off that in clangCodeGen.

**Review by Keenuts — APPROVED** [review]

**Pull request merged**